R/class-BipartiteNetwork.R

```diff
@@ -0,0 +1,50 @@
+check_bipartite_network <- function(object) {
+
+  errors <- character()
+
+  # Ensure that no node is in both columns
+  if (length(intersect(object@column1NodeIDs, object@column2NodeIDs)) > 0) {
+    errors <- c(errors, 'Bipartite networks cannot have nodes in both columns.')
+  }
+
+  # Check that all nodes are in at least one of the columns
+  if (!all(getNodeIds(object@nodes) %in% c(object@column1NodeIDs, object@column2NodeIDs))) {
+    errors <- c(errors, 'Found a node with a link that does not belong to either column. All nodes must exist in exactly one of the two columns in a bipartite network.')
+  }
+
+  return(if (length(errors) == 0) TRUE else errors)
+}
+
+#' Bipartite Network 
+#' 
+#' The bipartite network class represents data in the form of a network with two distinct groups of nodes
+#' in which nodes connect only with nodes from the other group. In other words, there are only inter-group
+#' links, no intra-group links. The two groups of nodes are commonly displayed as two columns of nodes.
+#' Bipartite networks can have any property of a regular network, but they also designate the node ids
+#' that belong to each column (group).
+#' 
+#' @slot links LinkList object defining the links in the network.
+#' @slot nodes NodeList object defining the nodes in the network. Some nodes may not have any links.
+#' @slot linkColorScheme string defining the type of coloring scheme the links follow. Options are 'none' (default) and 'posneg'.
+#' Use a method assignLinkColors() to assign colors to links and set this slot's value.
+#' @slot column1NodeIDs character vector listing the IDs of the nodes in the first column
+#' @slot column2NodeIDs character vector listing the IDs of the nodes in the second column
+#' 
+#' @name BipartiteNetwork-class
+#' @rdname BipartiteNetwork-class
+#' @include class-network.R
+#' @export 
+BipartiteNetwork <- setClass("BipartiteNetwork", 
+  contains = "Network",
+  representation(
+    column1NodeIDs = "character",
+    column2NodeIDs = "character"
```

Comment:
Good question. Because I think it's simpler from a data organization perspective to view the groups/columns as properties of the network. Like this thing is "a network with these extra properties". This organization is true to concept and also easy to expand upon when we add longitudinal and multilayer networks.

Reply:
maybe i'll make it a multipartite class instead of bipartite... That would mean we have something like a partitions slot that we can add lists of ids to. 
I still like having the nodes all together because the bipartite is more a partitioning with rules of a thing than a gluing together multiple things concept.

Reply:
Ok. I'm better able to get behind the partition slot I think and the more abstract multipartite concept.

Reply:
Can you say more about longitudinal and multilayer? Just for my benefit..

Reply:
For multilayer and longitudinal, we can think of the layer or timepoint as a property of the node and/or link. For multilayer, i'd also assign the nodes based on `layers` slot which would be a list of node id vectors. For temporal networks, it'll depend on the data but there's a more temporal version of multilayer networks we could take or make it more general and allow links to appear and disappear as well. In the latter case, we'd timestamp the nodes and links individually, and expand the slots to allow for capturing when a node or link is present.

Reply:
I think maybe what I want is to understand the difference between a partition and a layer and whether these things can coexist